### PR TITLE
print external LB address in admin kubeconfig

### PR DESCRIPTION
**Pull request by trawler**
Signed-off-by: Karen Almog <kalmog@mirantis.com>

**Issue**

**What this PR Includes**
When printing Admin's kubeconfig using `kubeconfig admin`, replace the localhost address with an external Address, if exists.

### CODE REVIEW

**Review by jnummelin — DISMISSED** [review]
{nitpick}
Bit unconventional to use `*string` as the return value, usually in errors funcs return just empty string with the error. 
{/nitpick}

File: cmd/kubeconfig.go [hunk 1]
```diff
@@ -93,12 +94,12 @@ Note: A certificate once signed cannot be revoked for a particular user`,
 				return errors.New("Username is mandatory")
 			}
 			var username = args[0]
-			clusterConfig, err := ConfigFromYaml(cfgFile)
-			if err != nil {
-				return err
-			}
 			var config = constant.GetConfig(dataDir)
 
+			clusterAPIAddress, err := getAPIAddress()
```
Comment (ncopa) [inline cmd/kubeconfig.go:99]:
```suggestion
			clusterAPIURL, err := getAPIURL()
```

File: cmd/kubeconfig.go [hunk 2]
```diff
@@ -136,7 +137,7 @@ Note: A certificate once signed cannot be revoked for a particular user`,
 				ClientCert: base64.StdEncoding.EncodeToString([]byte(userCert.Cert)),
 				ClientKey:  base64.StdEncoding.EncodeToString([]byte(userCert.Key)),
 				User:       username,
-				JoinURL:    clusterConfig.Spec.API.APIAddress(),
+				JoinURL:    *clusterAPIAddress,
```
Comment (ncopa) [inline cmd/kubeconfig.go]:
```suggestion
				JoinURL:    *clusterAPIURL,
```

File: cmd/kubeconfig.go [hunk 3]
```diff
@@ -163,11 +164,26 @@ Note: A certificate once signed cannot be revoked for a particular user`,
 				if err != nil {
 					log.Fatal(err)
 				}
-				os.Stdout.Write(content)
+
+				clusterAPIAddress, err := getAPIAddress()
```
Comment (ncopa) [inline cmd/kubeconfig.go:168]:
```suggestion
				clusterAPIURL, err := getAPIURL()
```

**Comment by trawler** [comment]
> {nitpick}
> Bit unconventional to use `*string` as the return value, usually in errors funcs return just empty string with the error.
> {/nitpick}

fixed

File: cmd/kubeconfig.go [hunk 4]
```diff
@@ -163,11 +164,26 @@ Note: A certificate once signed cannot be revoked for a particular user`,
...
+				if err != nil {
+					return errors.Wrap(err, "failed to fetch cluster's API Address: %v.")
+				}
+				newContent := strings.Replace(string(content), "https://localhost:6443", clusterAPIAddress, -1)
```
Comment (ncopa) [inline cmd/kubeconfig.go:172]:
```suggestion
				newContent := strings.Replace(string(content), "https://localhost:6443", clusterAPIURL, -1)
```

File: cmd/kubeconfig.go [hunk 5]
```diff
@@ -163,11 +164,26 @@ Note: A certificate once signed cannot be revoked for a particular user`,
...
+				os.Stdout.Write([]byte(newContent))
 			} else {
 				return errors.Errorf("failed to read admin config, is the control plane initialized on this node?")
 			}
 			return nil
 		},
 	}
 )
+
+func getAPIAddress() (string, error) {
```
Comment (ncopa) [inline cmd/kubeconfig.go:182]:
```suggestion
func getAPIURL() (string, error) {
```

File: cmd/kubeconfig.go [hunk 6]
```diff
@@ -163,11 +164,26 @@ Note: A certificate once signed cannot be revoked for a particular user`,
...
+	clusterConfig, err := ConfigFromYaml(cfgFile)
+	if err != nil {
+		return "", err
+	}
+	address := clusterConfig.Spec.API.APIAddress()
+	return address, nil
```
Comment (ncopa) [inline cmd/kubeconfig.go:188]:
```suggestion
	return clusterConfig.Spec.API.APIAddress(), nil
```

Or does that not work?

**Review by ncopa — CHANGES_REQUESTED** [review]
I think the term "address" is a bit misleading when it in really is an URL.

I think we should clean that up other places as well, but at least we should not add more confusion.

**Comment by trawler** [comment]
I added an APIAddressURL in this commit: https://github.com/k0sproject/k0s/tree/e72d12c92ab7c9ed34103398a3ec00c324720c59

Once merged, I'll rebased this PR.

**Review by ncopa — DISMISSED** [review]

**Pull request merged**